### ISSUE TITLE
ASNHC - 9. Parts

### ISSUE THREAD

**Issue by Jegelewicz** [ArctosDB/data-migration#174]
Part names are controlled by a code table. Part names used in the current system need to be mapped to part names in the Arctos part code table.

**Comment by Jegelewicz** [comment]
Fwd: ORA-20001: Invalid part name: heart (alcohol) (Mamm) ORA-06512: at "UAM.TRG_CMPD_SPECIMENPART", line 37 ORA-04088: error during execution of trigger 'UAM.TRG_CMPD_SPECIMENPART'

There are a few hundred of these from this AM. Looks like some perl script - it hasn't melted anything that I can see so I guess I don't much care, but it's obviously not very good at trapping errors and the data in eg https://arctos.database.museum/guid/ASNHC:Mamm:14675 don't look particularly believable - I don't have a clue if that's a broken script or just not particularly believable legacy data. Should we be concerned?

-D

@diannakrejsa  any idea what's up?

**Comment by diannakrejsa** [comment]
This morning I think @snknts was working on adding some preservation (lysis buffer, ethanol, etc.) info to parts that had been loaded lacking that information, as well as addressing some parts that were created erroneously as duplicates. I see a little of each in that record. Heart is supposed to be "heart (ethanol)" correct?

**Comment by Jegelewicz** [comment]
Parts should match the [parts name code table](https://arctos.database.museum/info/ctDocumentation.cfm?table=CTSPECIMEN_PART_NAME). If the part name doesn't exist in the table, I think it will not be added to the record. @dustymc can elaborate.

**Comment by campmlc** [comment]
Preservation can also be added as a part attribute instead.

**Comment by Jegelewicz** [comment]
BTW, I see that there is a character issue in the verbatim locality for this record - you may want to either remove the offending accent or use the UTF code so that it appears correctly.
![image](https://user-images.githubusercontent.com/5725767/71046996-59926a00-20f7-11ea-8fb0-6e799c58a48e.png)

**Comment by diannakrejsa** [comment]
We know about the parts code table, just a whoopsie in using alcohol instead of ethanol I think.

Joey and I can try the preservation as part attribute upload instead if that is easier than the script he wrote or if it's needed in the future. We'll talk about UTF code too (since I don't know what that means and how to upload it properly and he probably already does since I've heard him use the term..).

Can I ask you all to add **tissue (lysis buffer)** as a part that can be on Mammalogy db's? Right now it's just Herp.

**Comment by snknts** [comment]
The alcohol pres. method was me.  After updating 300 or so manually I wrote a thing to update the 3700 remaining... smarter not harder and such..
I noticed it wasn't saving non-frozen updates mid-run and changed the logic to match the buffer used for other parts in the same vial (instead of our initial box id lookup logic)

They should all be corrected now.. but I'll double check.

**Comment by Jegelewicz** [comment]
> Can I ask you all to add tissue (lysis buffer) as a part that can be on Mammalogy db's?

Done

> Joey and I can try the preservation as part attribute upload instead if that is easier than the script he wrote or if it's needed in the future.

At some point we will migrate everything to that method (I hope) so you can use either right now, but might be best if you just stick with one to be consistent internally. Working smarter is awesome! We just wanted to make sure you knew that some of it wasn't working because of mis-match with the code table.

> We'll talk about UTF code too (since I don't know what that means and how to upload it properly and he probably already does since I've heard him use the term..).

Just replace the offending character with the appropriate UTF code. I ususally just Google to find the code. Here's a [table for reference](http://www.fileformat.info/info/charset/UTF-8/list.htm).

**Comment by dustymc** [comment]
See also http://handbook.arctosdb.org/documentation/encoding.html

I'm more than fine with smarter as well, but it's difficult to deal with thousands of errors from something that's not designed to handle that. PLEASE talk to me before firing things like that off - worst case, it'll probably lead to a discussion about how Arctos could be better. Best case, I might be able to point you to existing tools...

<img width="327" alt="Screen Shot 2019-12-18 at 8 18 31 AM" src="https://user-images.githubusercontent.com/5720791/71103461-0956ee80-216f-11ea-9dc9-8ccd6091ed8c.png">

... that efficiently do what you need.

**Comment by snknts** [comment]
will do.  
I tried that tool.. but could seem to get the search to only return "heart"  (it included "heart (frozen)" too

**Comment by snknts** [comment]
*could not

(maybe typos like that is why.. sigh...)

**Comment by Jegelewicz** [comment]
Even so, you would then just modify all "heart" to "heart (ethanol)" which wouldn't affect the "heart (frozen)" parts...

**Comment by diannakrejsa** [comment]
Is there a way to express a partial skull in the parts code table? I'm talking about a crunched skull, not a cranium or mandible, and not a post-cranial "skeleton, partial". 

Can we add "skull, partial"?

**Comment by Jegelewicz** [comment]
> Can we add "skull, partial"?

I dislike it very much! Here is what I do. Add the part as skull, condition as "partial, crunched".

The skeleton, partial is problematic because it is a mixture of part and condition, but it is already there. At some point, I will suggest we lose that and just use skeleton with "partial" in condition.

**Comment by dustymc** [comment]
Yes getting all modifiers out of part name is on our hit-list; please don't add more! "Partial" and "smooshed" and such (the more precise and descriptive the better) definitely belong in condition.

**Comment by diannakrejsa** [comment]
Noted! We'll be sure to use smooshed on condition for something soon.

**Comment by diannakrejsa** [comment]
May I request an existing code table part be added to Mamm databases?
**heart, liver (95% ethanol)** is currently only allowed in Herp databases.
Thanks!

**Comment by Jegelewicz** [comment]
Done!

**Comment by diannakrejsa** [comment]
Another parts question, where do we log whether a nobuto strip was taken for a record? Is that **blood (dry)**?

**Comment by Jegelewicz** [comment]
Here's how DMNS:Mamm handles it

![image](https://user-images.githubusercontent.com/5725767/73949228-5de93100-48b7-11ea-8261-2c103e4dce6b.png)

If they aren't frozen, then using blood (dry) with Nobuto in remark seems like a good option. Would "nabuto strip" be an appropriate preservation method? Perhaps we should add "nabuto strip" to the preservation code table.

**Comment by Jegelewicz** [comment]
Yep here's one from MSB

![image](https://user-images.githubusercontent.com/5725767/73949513-bb7d7d80-48b7-11ea-8889-3cb58b3c5fc3.png)

**Comment by diannakrejsa** [comment]
Yes perhaps the nobuto strip in preservation would be good. I can also add to remarks like everyone else has been doing, it lets you know where to look within your own collection for the sample (among the nobutos). Whatever seems best for organization & simplicity.

**Comment by Jegelewicz** [comment]
For now, use the remark so everyone will be consistent. I'll propose the new preservation term and eventually we can migrate everyone to that format.

**Comment by dustymc** [comment]
As always, the "correct" answer (which we may or may not be able to do anything about!) comes from "who cares?"

1. Nobody who hasn't already found the part===>remarks is fine.
2. Somebody might want to search===>free-text can't do that.

It took me about 5 seconds to find this:

```
Created in SPECIFY; 2014-06-07|Last modified by SPECIFY; 2014-06-07|Nobunto strip taken
```

I could spend the rest of the week, and still not REALLY know if I'd found them all.

**Comment by diannakrejsa** [comment]
Ugh, yes controlled vocabulary/drop downs are so useful...

**Comment by diannakrejsa** [comment]
May I request **blood**, **kidney**, and **muscle** be added as part names for Bird databases?

Thanks!

**Comment by Jegelewicz** [comment]
Added.
